### DISCUSSION TITLE
`ValidateAppProtocolVersion` should not check extra data

### DISCUSSION THREAD

**Discussion by Atralupus** [planetarium/libplanet#2367] [General]
We implemented a separate updater in the launcher for each project(player, launcher...) - https://github.com/planetarium/9c-launcher/pull/1731

It's almost done, and this updater is dependent on `AppProtocolVersion` extra data - https://github.com/planetarium/9c-launcher/issues/1744
`AppProtocolVersion` Shema looks like below
```
APV(100291)
player: "1/3f26a4c7a2f4fecbca8fe45a5969b85b5b3a189a"
launcher: "1/39aa07191148615acdb6dd48e9d05c7446f3d6b7"
```
Now, if only the player changes exist, then only the player part will be changed. not increase `AppProtocolVersion` version

But Currently `ValidateAppProtocolVersion` method in `MessageValidator` is check `AppProtocolVersion` equality using this code
https://github.com/planetarium/libplanet/blob/d05c8434a1682a761e45eebc3eb991a2375ea9f9/Libplanet.Net/AppProtocolVersion.cs#L239
We want the `ValidateAppProtocolVersion` method should not to check extra data

**Comment by Atralupus** [comment]
I tried to solve this issue in https://github.com/Atralupus/libplanet/tree/feat/except-extra-camparer this branch
but I've faced [Sign](https://github.com/planetarium/libplanet/blob/1a41a33750ab306c4b67346d52415f72b222b444/Libplanet.Net/AppProtocolVersion.cs#L123) method, Sign using extra data for creating signature

**Comment by dahlia** [comment]
Although Nine Chronicles wants Libplanet to consider any APVs signed by trusted parties even if their extra data differ from the current node's one, it's still unsure if it's generally good practice for every game/app made with Libplanet.  IMHO it should be *configurable*.

In fact, there was an option to accept or ignore different APVs signed by trusted parties, but it was removed for the simplicity (see also: <https://github.com/planetarium/libplanet/issues/1885>).  It might be better to revert the removal (or provide a better API for it).

**Reply by Atralupus** [reply]
That Issue seems to exist with the option to accept different APV, not check APV's extra part.
But, This issue wants to check extra data specifically
![image](https://user-images.githubusercontent.com/30599098/194002278-9e8e7d46-e68b-47da-b426-1227cbc9a8f2.png)


As you mentioned, I would provide a better API to check extra
Also, I will put effort into this function can be configurable

**Reply by greymistcube** [reply]
> Although Nine Chronicles wants Libplanet to consider any APVs signed by trusted parties even if their extra data differ from the current node's one, it's still unsure if it's generally good practice for every game/app made with Libplanet. IMHO it should be configurable.

I don't quite agree with this sentiment. Although the intention of APVs as they are used by us may be a bit more clear to us, I'd say providing an option what to do with it makes what APVs are for even more vague. To an outsider, it may look like "Here is this **thing** called APV, do whatever you want with it". At that point, we might as well off-load APV validation and handling to the application and use something like `IsAppProtocolVersionAcceptable` and/or `AppProtocolVersionHandler` delegates. 😶

Also the removed option is too "coarse" for the issue at hand as @Atralupus points out. See the comment below. 😗

**Comment by greymistcube** [comment]
I was a bit confused for a while and had to go over the context several times to understand what is going on (I think?). 🙄

## Current Implementation

APV is structured to have three parts:

- Version number
- Signer identity
- Extra data

As it stands, APV equality is defined as "exact equality" where all three of the above must match. If a node receives a message with a different APV, two things "happen":

- The receiving node triggers the `DifferentAppProtocolVersionEncountered` delegate provided by the application, namely [NineChronicles] for our case, in order to let the application handle whatever that is needed to be done.
- [Libplanet] itself **does nothing** with the received message with a different APV.

## Intention of APV

Although **not specified explicitly** in `AppProtocolVersion`, I think it'd be safe to assume "different" APV, as in **different version number**, i.e. `Version`, not as in not equal as defined above, is supposed to mean the nodes would not be compatible due to how [Libplanet] is designed[^1]. Essentially, a higher APV signals that for the receiving node execution results are (possibly) going to be different from the node sending out the message. Under normal circumstances, we'd want the lower number APV node to trigger auto-update the client so that it can properly sync up to the nodes with a higher APV.

## Wanted Feature

From the application side, although I'm not directly involved, I'm guessing the reason for the original post and the diagram by @Atralupus is to have launcher update cycle decoupled with the player update cycle. This can be interpreted several ways:

- Trigger launcher update without full deployment with an APV bump.
- Have launcher update while the player is **continuously running**.[^2]

In either case, some kind of callback to the external application must be triggered. As for game clients, if we go with the first option, this isn't much of an issue as `DifferentAppProtocolVersionEncountered` will be triggered as it stands, and "distributed nodes" for players can be forcefully updated if needed.

However, the problem is further complicated by [NineChronicles.Headless] nodes, as these are independent from the launcher client, as we'd want, or at least that is how I understand it, the deployed cluster intact.

## Premise

I think it'd be a good idea to define basic premises for APVs to make this discussion more concise. Let's define some terms first (by borrowing from math).

- Two APVs are said to be **equal** if `Version`s, `Signer`s, and `Extra`s are all the same.
- Two APVs are said to be **strongly equivalent** if `Version`s and `Signer`s are the same.[^3]
- Two APVs are said to be **weakly equivalent** if `Version`s are the same.

Consider the following premise:

- Any two nodes with **weakly equivalent** APVs are compatible.

If we accept this as part of the spec for [Libplanet], then I'd say it makes sense for any node to accept all messages with **weakly equivalent** APVs as this should guarantee proper consensus between nodes.[^4]

Then the issue at hand seems to be that the calling conditions for `DifferentAppProtocolVersion` and message accepting conditions are treated as the same when they should not be.



[^1]: The library depending on externally provided assembly to execute game actions.
[^2]: I'm not sure how this would be possible as a client's backend would have to use the updated APV.
[^3]: Although this hasn't come into play yet really, but it could be used as a fail-safe mechanism. Other than that, in order to truly trust a different signer, we'd need an air-tight specification of [Libplanet], whatever application that might be using [Libplanet], possibly a test suite to pass. I don't know how realistic this scenario may be in any foreseeable future. 🙄
[^4]: The validity of a received APV is an entirely different issue. This must be checked separately.



[Libplanet]: https://github.com/planetarium/libplanet
[NineChronicles]: https://github.com/planetarium/NineChronicles.
[NineChronicles.Headless]: https://github.com/planetarium/NineChronicles.Headless

**Comment by Atralupus** [comment]
We discussed and draw out the results
Just append only check APV version https://github.com/planetarium/libplanet/blob/71e204f4dc9bc4ee27e55122cf0eaccb52d7cff7/Libplanet.Net/Messages/MessageValidator.cs#L157 above this line

Currently `AppProtocolVersion` values `Version` not other data (FYI @greymistcube 's `Intention of APV`)

**Comment by riemannulus** [comment]
I guess this discussion is still available and AFAIK @Atralupus and @greymistcube are discussed on discord about this. Have you any issues or ideas about this discussion?
